lib/src/http.dart

```diff
@@ -343,58 +231,202 @@ Map parseJsonResponse(http.Response response) {
 Never invalidServerResponse(http.Response response) =>
     fail(log.red('Invalid server response:\n${response.body}'));
 
+/// Exception thrown when an HTTP operation fails.
```

Comment:
You now have two exceptions... PubHttpException and this... Documentation comments don't explain why.

Reply:
So, `PubConnectException` was added as a temporary hack because `PubHttpException` requires a non-null "response" as part of its construction. This became an issue when I was trying to wrap `SocketException`s. When these happen, it's early enough within the request lifecycle that there is no "response" object I could use to create a `PubHttpException`.

#### What I ended up doing:
- Create `PubConnectException`, which extends `WrappedException`, with the intention to wrap `IOException`, `SocketException`, and possibly `TimeoutException` and `AuthenticationException`.

#### Why wrap?
Initially, I did it because it enables us to have a centralized `couldRetry` property and to drive down the complexity of the `retryIf` function. This is through `mapException`, which I'm using to parse and wrap exceptions before `retryIf` is called.

Thinking further on it, I think it could also improve logs and make it easier to do metrics for HTTP failures.

#### Some options I've been thinking about:
1. Don't wrap exceptions and remove `PubConnectException`.
    - Simplest option and possibly the best one?
2. Make "response" optional in `PubHttpException`.
    - This means updating some logging code to have null checks.
    - Make `PubHttpException` conform to `WrappedException`?
      - Use `PubHttpException` to wrap `IOException`, `SocketException`, and possibly `TimeoutException` and `AuthenticationException`.
3. Rename `PubHttpException` to `PubHttpResponseException` and rename `PubConnectException` to `PubHttpConnect(ion?)Exception`.
    - Use `PubHttpConnect(ion?)Exception` to wrap `IOException`, `SocketException`, and possibly `TimeoutException` and `AuthenticationException`.

Reply:
> Don't wrap exceptions and remove PubConnectException.
>  Simplest option and possibly the best one?

Yes, wrapping exceptions is always extremely complicated.

> Make "response" optional in PubHttpException.

Or maybe we have two:
```dart

class PubHttpException {
  final bool isIntermittent;
  final String message;
}

class PubHttpResponseException extends PubHttpException {
  final int status;
  final Map<String,String> headers;
}

```

Reply:
The example code you have feels like a good option. I'll work on adopting it and then see how it feels overall.